### PR TITLE
Documented the Visualizer architecture

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Added proper documentation for the Visualizer architecture and the basics of how it is set up and edited with Qt.

### Testing Done

N/A

### Resolved Issues

#1087 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
Tagging lots of people both because I don't know who's free over the break, and because we want to make sure this is useful to everyone.

File: docs/architecture-and-design.md [hunk 1]
```diff
@@ -387,10 +392,60 @@ Most [Intents](#intents) are easy to break down into  [Primitives](#primitives),
...
 
-The `Visualizer` also lets us control the [AI](#ai) by setting [Dynamic Parameters](#dynamic-parameters). Through the `Visualizer`, we can manually choose what strategy the [AI](#ai) should use, what colour we are playing as (yellow or blue), and tune more granular behaviour such as how close an enemy must be to the ball before we consider them a threat.
+The [Visualizer](#visualizer) also lets us control the [AI](#ai) by setting [Dynamic Parameters](#dynamic-parameters). Through the [Visualizer](#visualizer), we can manually choose what strategy the [AI](#ai) should use, what colour we are playing as (yellow or blue), and tune more granular behaviour such as how close an enemy must be to the ball before we consider them a threat.
+
+The [Visualizer](#visualizer) is connected to the rest of the system using the [Observer Design Pattern](#observer-design-pattern). It observes Subjects that contain information it wants to display, such as the [World(#world) or [DrawFunctions](#draw-functions).
```
Comment (garethellis0) [inline docs/architecture-and-design.md]:
``` suggestion
The [Visualizer](#visualizer) is connected to the rest of the system using the [Observer Design Pattern](#observer-design-pattern). It observes Subjects that contain information it wants to display, such as the [World](#world) or [DrawFunctions](#draw-functions).
```

File: docs/architecture-and-design.md [hunk 2]
```diff
@@ -387,10 +392,60 @@ Most [Intents](#intents) are easy to break down into  [Primitives](#primitives),
...
+
+The [Visualizer](#visualizer) is implemented using [Qt](https://www.qt.io/), a C++ library for creating cross-platform GUIs. The general documentation for [Qt](https://www.qt.io/) can be found [here](https://doc.qt.io/qt-5/index.html). The most important fundamentals are:
```
Comment (garethellis0) [inline docs/architecture-and-design.md]:
``` suggestion
The [Visualizer](#visualizer) is implemented using [Qt](https://www.qt.io/), a C++ library for creating cross-platform GUIs. The general documentation for [Qt](https://www.qt.io/) can be found [here](https://doc.qt.io/qt-5/index.html). The most important parts for the Visualizer are:
```

File: docs/architecture-and-design.md [hunk 3]
```diff
@@ -387,10 +392,60 @@ Most [Intents](#intents) are easy to break down into  [Primitives](#primitives),
...
+* [Signals and Slots](https://doc.qt.io/qt-5/signalsandslots.html)
+* [QtCreator](https://doc.qt.io/qtcreator/creator-using-qt-designer.html) (specifically for Widget-based applications)
+* [Widgets](https://doc.qt.io/qt-5/qtwidgets-index.html)
+
+The [Visualizer](#visualizer) is made up of 2 major components: The [QApplication](https://doc.qt.io/qt-5/qapplication.html) and `Visualizer Widget` that contains all of the graphical components used in the [Visualizer](#visualizer), and the `VisualizerWrapper`. The `VisualizerWrapper` runs the [QApplication](https://doc.qt.io/qt-5/qapplication.html) in a separate thread, so that Qt can run its event loop and handle events and rendering without blocking our main thread.
```
Comment (garethellis0) [inline docs/architecture-and-design.md]: This reads like there are more then two components (`QApplication`, `Visualizer Widget` and `VisualizerWrapper`). A two-long bullet-point list might be a bit better here.

Comment (garethellis0) [inline docs/architecture-and-design.md]: Assuming I understand this correct, the `VisualizerWrapper` contains an instance of the `VisualizerWidget`. If that's the case, describing these as "two major components" is a bit confusing to me, in that it made me think of them as being separate.

File: docs/architecture-and-design.md [hunk 4]
```diff
@@ -387,10 +392,60 @@ Most [Intents](#intents) are easy to break down into  [Primitives](#primitives),
...
+
+## Visualizer Diagram
+![Visualizer Diagram](images/visualizer_diagram.svg)
+
+## Inter-thread Communication
+The `VisualizerWrapper` needs to communicate with the [QApplication](https://doc.qt.io/qt-5/qapplication.html) and `Visualizer Widget` running in its separate thread in order to trigger events like drawing when new data is received. In order to do this, the `VisualizerWrapper` and `Visualizer Widget` use our `ThreadsafeBuffer` class to communicate. The `VisualizerWrapper` pushes data into the buffers, and the `Visualizer Widget` pops the data in a `Producer -> Consumer` pattern. This means the `Visualizer Widget` can handle data at its own rate, indepdendant from the `VisualizerWrapper`.
```
Comment (garethellis0) [inline docs/architecture-and-design.md]: `indepdendant` -> `independent`

File: docs/architecture-and-design.md [hunk 5]
```diff
@@ -387,10 +392,60 @@ Most [Intents](#intents) are easy to break down into  [Primitives](#primitives),
...
+For a very quick tutorial on how to use QtCreator, see [this video](https://www.youtube.com/watch?v=R6zWLfHIYJw)
+
+To summarize, [QtCreator](https://doc.qt.io/qtcreator/creator-using-qt-designer.html) creates and modifies a `.ui` file, which is more-or-less an `XML` describing the GUI application (what components exist, how they are positioned relative to one another, and their attributes). During compilation, this `.ui` file gets generated into code which handles all the setup and layout of the GUI components that have been defined in the `.ui` file. We include the autogenerated code in our [Visualizer](#visualizer) code where we are then able to connect the autogenerated widgets to various functions, and implement the logic we need to.
+
+### Editing `.ui` files
+1. Open QtCreator
+2. Click `File -> Open File or Project`
+3. Select the `.ui` file.
+4. Make your changes
+  * Don't forget to save. You must save the file for changes to be picked up in compilation
```
Comment (garethellis0) [inline docs/architecture-and-design.md]: formatting suggestion:
``` 
4. Make your changes (*Don't forget to save. You must save the file for changes to be picked up in compilation*)
```

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Nothing major, small formatting/phrasing comments.

**Review by EvanMorcom — DISMISSED** [review]
LGTM pending Gareths comments :100:

**Review by henryymliu — DISMISSED** [review]
Nothing to add

File: docs/architecture-and-design.md (hunk 4 above)
Comment (MathewMacDougall) [reply docs/architecture-and-design.md]: fixed

File: docs/architecture-and-design.md (hunk 5 above)
Comment (MathewMacDougall) [reply docs/architecture-and-design.md]: changed as suggested

File: docs/architecture-and-design.md (hunk 3 above)
Comment (MathewMacDougall) [reply docs/architecture-and-design.md]: See above reply.

Comment (MathewMacDougall) [reply docs/architecture-and-design.md]: Yup I agree this sentence / paragraph wasn't clear, and I talk about the wrapper too despite saying there's only 2 major components. I have separated this into a bullet list (and differentiated Qt and non-Qt components) which should be much more clear.

File: docs/architecture-and-design.md [hunk 6]
```diff
@@ -387,10 +392,64 @@ Most [Intents](#intents) are easy to break down into  [Primitives](#primitives),
...
+  * The `Visualizer Widget`. This contains all of the graphical components used in the [Visualizer](#visualizer).
+* Non-Qt Components
+  * The `VisualizerWrapper`. The `VisualizerWrapper` contains the [QApplication](https://doc.qt.io/qt-5/qapplication.html) and `Visualizer Widget`. It runs the [QApplication](https://doc.qt.io/qt-5/qapplication.html) in a separate thread, so that Qt can run its event loop and handle events and rendering without blocking our main thread.
+ 
+
+## Visualizer Diagram
+![Visualizer Diagram](images/visualizer_diagram.svg)
+
+## Inter-thread Communication
+The `VisualizerWrapper` needs to communicate with the [QApplication](https://doc.qt.io/qt-5/qapplication.html) and `Visualizer Widget` running in its separate thread in order to trigger events like drawing when new data is received. In order to do this, the `VisualizerWrapper` and `Visualizer Widget` use our `ThreadsafeBuffer` class to communicate. The `VisualizerWrapper` pushes data into the buffers, and the `Visualizer Widget` pops the data in a `Producer -> Consumer` pattern. This means the `Visualizer Widget` can handle data at its own rate, indepdendent from the `VisualizerWrapper`.
```
Comment (garethellis0) [inline docs/architecture-and-design.md]: "indepdendent" -> "independent"

**Comment by garethellis0** [comment]
One small spelling fix I missed first time through, but LGTM otherwise!

File: docs/architecture-and-design.md (hunk 6 above)
Comment (MathewMacDougall) [reply docs/architecture-and-design.md]: fixed

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada: - pending others review, this definitely needs a few eyeballs on it

**Review by jonathanlew — APPROVED** [review]
Should be helpful for making mrftest. Thanks for this!

**Pull request merged**